### PR TITLE
Absorbtion for isis direct inelastic scripts

### LINKED ISSUE #26871: Bring Absorbtion corrections into Direct inelastic workflow.
Direct inelastic group in ISIS requests incorporating absorbtion corrections in the inelastic data reduction workflow. 

It looks like the best way of achieving this would be adding "correct_absorbtion_on" property to direct inelastic reduction properties and implement range of Python classes, which describe different shapes and have the property, which deploys appropriate Mantid Absorbtion correction algorithm. 

The corrections then would be deployed by adding something like:
absorbtion_correct_on = AbsrbCylinder(r=,h=,axis = [1,0,0]))
absorbtion_use_MonteCarlo = True(False)

### PR DESCRIPTION
**Pull request by abuts**
**Description of work.**
This brings existing absorption corrections algorithms within direct inelastic workflow as requested by the group.
To achieve this, I have written number of python classes which describe common shapes and two classes, which bring absorption corrections into direct inelastic workflow

**To test:**
As the work relies on existing functionality, a code review would probably be sufficient. The real test of success or failure of this work would be feedback from the instrument scientists and users 

Fixes #26871 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/test/AbsorptionShapesTest.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        self.assertEqual(res['Thick'],2)
+        self.assertEqual(res['Center'],[0,0,0])
+        self.assertEqual(res['Angle'],20)
+
+        test_ws = CreateSampleWorkspace(NumBanks=1,BankPixelWidth=1)
+        test_ws = ConvertUnits(test_ws,'DeltaE',Emode='Direct',EFixed=2000)
+        #cor_ws,corrections = ash.correct_absorption(test_ws,ElementSize=5)
+        #n_bins = corrections.blocksize()
+        #corr_ranges = [n_bins,corrections.readY(0)[0],corrections.readY(0)[n_bins-1]]
+        #np.testing.assert_almost_equal(corr_ranges,[97,0.4504,0.0009],4)
```
Comment (mducle) [inline scripts/test/AbsorptionShapesTest.py]: Should these commented out lines be removed? I guess they're using the "fast" algorithm? Why not uncomment them and let them run?

File: scripts/test/AbsorptionShapesTest.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        self.assertEqual(res['Axis'],[0,1,0])
+        self.assertEqual(res['Center'],[0,0,0])
+
+
+        test_ws =  CreateSampleWorkspace(NumBanks=1,BankPixelWidth=1)
+        test_ws = ConvertUnits(test_ws,'DeltaE',Emode='Direct',EFixed=2000)
+        #cor_ws,corrections = ash.correct_absorption(test_ws,ElementSize=5)
+        #n_bins = corrections.blocksize()
+        #corr_ranges = [n_bins,corrections.readY(0)[0],corrections.readY(0)[n_bins-1]]
+        #np.testing.assert_almost_equal(corr_ranges,[97,0.2984,0.0002],4)
```
Comment (mducle) [inline scripts/test/AbsorptionShapesTest.py]: Should these commented out lines be removed?

File: scripts/test/AbsorptionShapesTest.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        test_ws = ConvertUnits(test_ws,'DeltaE',Emode='Direct',EFixed=2000)
+        #cor_ws,corrections = ash.correct_absorption(test_ws,ElementSize=6)
+        #n_bins = corrections.blocksize()
+        #corr_ranges = [n_bins,corrections.readY(0)[0],corrections.readY(0)[n_bins-1]]
+        #np.testing.assert_almost_equal(corr_ranges,[97,0.6704,0.4097],4)
+
+        ##cor_ws,corrections = ash.correct_absorption(test_ws)
+        #n_bins = corrections.blocksize()
+        #corr_ranges = [n_bins,corrections.readY(0)[0],corrections.readY(0)[n_bins-1]]
+        #np.testing.assert_almost_equal(corr_ranges,[97,0.6420,0.3997],4)
```
Comment (mducle) [inline scripts/test/AbsorptionShapesTest.py]: Should these commented out lines be removed?

**Review by mducle — CHANGES_REQUESTED** [review]
Mostly spelling mistakes but there are some bits of the logic I didn't get...

File: Testing/SystemTests/tests/analysis/ISIS_LETReduction.py [hunk 4]
```diff
@@ -249,7 +297,65 @@ def custom_name(prop_man):
...
+
+            Input:
+            If provided, the pointer or the name of the workspace available in analysis data service.
+            If it is not, the workspace is taken from PropertyManager.sample_run property
+
+            Usage:
+            Reduce single run and uncomment this method in the __main__ area to evaluate
+            adsorption corrections.
 
+            Change adsorption corrections parameters below to achieve best speed and
```
[the remark this answers is not in the document]
Comment (abuts) [reply Testing/SystemTests/tests/analysis/ISIS_LETReduction.py]: Done

**Comment by abuts** [comment]
Good remarks, I hope all addressed in number of recent commits. -- Have problems in commentig one by one in my browser

File: Testing/SystemTests/tests/analysis/ISIS_MERLINReduction.py [hunk 5]
```diff
@@ -55,33 +59,161 @@ def reduce(self,input_file=None,output_directory=None):
         outWS = ReductionWrapper.reduce(self,input_file,output_directory)
      #SaveNexus(ws,Filename = 'MARNewReduction.nxs')
         return outWS
+      #
```
[the remark this answers is not in the document]
Comment (mducle) [reply Testing/SystemTests/tests/analysis/ISIS_MERLINReduction.py]: Discussed with AB, and decided to keep additional stubs in each instrument reduction file as is for documentation purposes.

**Review by mducle — APPROVED** [review]
Looks good to me now!

**Pull request merged**